ref(hc):  RPC in tests uses test client and hits rpc endpoint

corps opened the pull request:
RPC Services use the django test client in tests rather than falling back to local service.
Several refactors and lots of code removal related to this in addition to fixing several service issues.

corps committed:
ref(hc):  RPC in tests uses test client and hits rpc endpoint

corps committed:
Demo it working!!!

corps commented:
So, as is, this works!  Run `hybrid_cloud/test_organization.py` to see this working end to end!

But it requires us to find some solutions to these problems:

1.  Conditionally swap backend out between sending test requests
2.  Fix remaining signature issues
3.  Measure performance and see how to have this exercise code without it being too much impact to CI

corps commented:
Another complication I forgot -- we actually DO need to keep stubbed around for production! Duh.


I'd like to test with the test http client though as much as possible.  So we need 3 backends

1.  Pass through backend, that acts like stubbed
2.  Test client backend, that acts like stubbed BUT actually issues fake http request
3.  Real backend

corps committed:
Merge branch 'master' into zc/rpc-in-tests

corps committed:
More fixes

corps committed:
Fixes

corps committed:
More fixes

corps committed:
Fixing test decorato

corps committed:
More fixes

RyanSkonnord commented on src/sentry/services/hybrid_cloud/rpc.py:
```diff
@@ -466,59 +444,115 @@ def result_to_dict(value: Any) -> Any:
+        if response.status_code != 200:
+            if in_test_environment():
+                if response.status_code == 500:
+                    raise Exception(
+                        "Error invoking rpc at {path}: check error logs for more details"
+                    )
+                raise Exception("Error invoking rpc at {path}: {response.json()['detail']}")
```
The `f` is missing from the beginning of the strings.

Also, nit: adding an `!r` to the interpolations would make them a bit more readable (as in `{path!r}`, which is equivalent to `{repr(path)}`).

RyanSkonnord commented on src/sentry/services/hybrid_cloud/rpc.py:
```diff
@@ -319,31 +320,27 @@ def _create_signatures(cls) -> Mapping[str, RpcMethodSignature]:
-                        f"Error on parameter model for {cls.__name__}.{base_method.__name__}: {e}"
-                    )
+                raise RpcServiceSetupException(
+                    f"Error on parameter model for {cls.__name__}.{base_method.__name__}: {e}"
```
Should be safe to drop the `": {e}"` from the end, as we now have the `from e`.

corps commented on src/sentry/services/hybrid_cloud/rpc.py:
```diff
@@ -466,59 +444,115 @@ def result_to_dict(value: Any) -> Any:
-        None
-        if return_value is None
-        else service.deserialize_rpc_response(method_name, return_value)
+        if response.status_code != 200:
+            if in_test_environment():
+                if response.status_code == 500:
+                    raise Exception(
+                        "Error invoking rpc at {path}: check error logs for more details"
+                    )
+                raise Exception("Error invoking rpc at {path}: {response.json()['detail']}")
+            # Careful not to reveal too much information in production
+            if response.status_code == 403:
+                raise Exception("Unauthorized service access")
+            if response.status_code == 400:
+                raise Exception("Invalid service request")
+            raise Exception("Service unavailable")
+        serial_response = response.json()
+    return serial_response
```
Sure!  I can make that change.

corps committed:
Merge branch 'master' into zc/rpc-in-tests

RyanSkonnord commented:
The overall approach seems solid, and the added complexity seems valid.

It seemed like `_dispatch_to_silo_service` was getting a bit big and blocky, so I took a pass at chunking it out and pulling some common values to the object level in https://github.com/getsentry/sentry/pull/53305

corps committed:
Removing unused monolith conditional path

RyanSkonnord committed:
ref: Refactor _dispatch_to_silo_service for readability (#53305)

Suggested changes for https://github.com/getsentry/sentry/pull/53237

corps committed:
Fixes

corps committed:
Capture errors from serialization in tests

RyanSkonnord committed:
Restore dispatch_remote_call as a hook for mocks

RyanSkonnord committed:
Make AuditLogEvent.actor_label nullable to match DB model

RyanSkonnord committed:
Merge branch 'master' into zc/rpc-in-tests

RyanSkonnord committed:
Fix unserialized user passed to user_service.serialize_many in get_client_config

RyanSkonnord committed:
Introduce custom RpcRemoteException type for _raise_from_response_status_error

RyanSkonnord committed:
Fix _RemoteSiloCall attribute order (oops 😩)

RyanSkonnord committed:
Docstring edit

RyanSkonnord committed:
Make RpcRepository.integration_id nullable to match DB model

RyanSkonnord committed:
Break up RpcSendException checks for finer-grained messages

RyanSkonnord committed:
Remove RPC_SHARED_SECRET check, because it's redundant to other checks closer to where it's used

RyanSkonnord committed:
Replace RpcSendException with RpcServiceSetupException

RyanSkonnord committed:
Set RPC_SHARED_SECRET in other threads in OutboxDrainTest

RyanSkonnord committed:
Make RpcNotificationSetting.target_id nullable to match DB model

RyanSkonnord committed:
Update reference to RpcSendException in test

RyanSkonnord committed:
Override in_test_environment in test_models_tolerate_invalid_types

Move the `in_test_environment()` check in
`_hack_pydantic_type_validation` to inside the callback. Has the nice
side effect that we exercise the delegation to `builtin_validate`, which
we rely on in prod.

RyanSkonnord committed:
In TransformTestCase, persist NotificationSettings to populate their `id` attr

RyanSkonnord committed:
Add RPC_SHARED_SECRET to settings on MsTeamsProxyApiClientTest

RyanSkonnord committed:
Better style for TransformTestCase fix

RyanSkonnord committed:
Make RpcServiceHook.application_id nullable to match DB model

RyanSkonnord committed:
Add RPC_SHARED_SECRET override to test_find_regions_for_user

RyanSkonnord committed:
Set RPC_SHARED_SECRET in global test settings; revert piecemeal overrides

RyanSkonnord committed:
Merge branch 'master' into zc/rpc-in-tests

RyanSkonnord committed:
Handle both RpcUser and User in get_client_config

RyanSkonnord committed:
Fix service interface glitches that were breaking OrganizationAuthSettingsTest

RyanSkonnord committed:
Fix RPC type-handling errors around incident status code

RyanSkonnord committed:
Fix bad argument to user_service.serialize_many

RyanSkonnord committed:
Make RpcSentryAppEventData serializable

RyanSkonnord committed:
Fix region setup in ProcessControlOutboxTest

RyanSkonnord committed:
Fix serialization bugs arising from GroupSerializerTest

RyanSkonnord commented on src/sentry/services/hybrid_cloud/user/serial.py:
```diff
@@ -1,13 +1,27 @@
 from sentry.services.hybrid_cloud.user import RpcAuthenticator, RpcAvatar, RpcUser, RpcUserEmail
 
 
+def serialize_generic_user(user: Any) -> Optional[RpcUser]:
+    """Serialize a user-representing object of unknown type to an RpcUser.
+
+    Return None if the user is anonymous (not logged in).
+    """
+    if user is None or user.id is None:
+        return None
+    if isinstance(user, RpcUser):
+        return user
+    if isinstance(user, User):
+        return serialize_rpc_user(user)
+    raise TypeError(f"Can't serialize {type(user)} to RpcUser")
```
Are we cool with this existing?

I'd really really rather live in a world where we can just rely on well-defined (and annotated) types. But I kept running into cases (especially around serialization) where a function just has an unannotated argument named `user` and it's floating around in a non-silo-scoped context.

corps replied:
Yeah, I can see the appeal.

RyanSkonnord committed:
Fix null filter args in RuleSerializer

RyanSkonnord committed:
Pass an actual list to integration_service from has_alert_integration

RyanSkonnord committed:
Pass an actual list to user_service from get_owners

RyanSkonnord commented:
One thing that's stumping me currently: some tests (including `ProxyTestCase` and `VerifyRequestBodyTest`) reports

> `django.urls.exceptions.NoReverseMatch: Reverse for 'sentry-api-0-rpc-service' not found. 'sentry-api-0-rpc-service' is not a valid view function or pattern name.`

despite the same `reverse` call working in other tests. The `responses` module seems to be the common factor but I'm not sure how to work around it.

RyanSkonnord committed:
Prevent passing None to user_service.get_user from GroupActivityNotification

RyanSkonnord committed:
Pass an actual list to user_service from get_participants_for_release

corps commented:
> One thing that's stumping me currently: some tests (including `ProxyTestCase` and `VerifyRequestBodyTest`) reports
> 
> > `django.urls.exceptions.NoReverseMatch: Reverse for 'sentry-api-0-rpc-service' not found. 'sentry-api-0-rpc-service' is not a valid view function or pattern name.`
> 
> despite the same `reverse` call working in other tests. The `responses` module seems to be the common factor but I'm not sure how to work around it.

This might be related to `override_settings`.  You need to make sure the application ends up with `ROOT_URLCONF="sentry.web.urls"`.  Some reasons why this may not be the case:

1.  The base test class doesn't include this decorator.
2.   The class decorator's values aren't being captured in subclasses
3.  various override_settings bugs.

I'd try various methods to validate if forcing this value fixes the issue in the tests in question.

corps committed:
Merge branch 'master' into zc/rpc-in-tests

RyanSkonnord committed:
Merge branch 'master' into zc/rpc-in-tests

corps committed:
Merge branch 'master' into zc/rpc-in-tests

corps committed:
Using typing_extensions.TypedDict

RyanSkonnord committed:
Fix type token reading error caused by "from __future__ import annotations"

corps committed:
Minor fixes

RyanSkonnord committed:
Merge branch 'master' into zc/rpc-in-tests

RyanSkonnord committed:
Merge branch 'master' into zc/rpc-in-tests

RyanSkonnord committed:
Handle null value for UserSocialAuth.extra_data when serializing

Also make a defensive copy of the mutable dict, just in case.

RyanSkonnord committed:
Serialize auth token in get_client_config

RyanSkonnord committed:
Make sure to serialize user in IncidentActivitySerializer

RyanSkonnord committed:
Add type check to FilterQueryDatabaseImpl.serialize_many for convenience

corps committed:
Fixing

corps committed:
Conditionalize

corps committed:
More fixes

RyanSkonnord committed:
Fix "type: ignore" on `**extra` arg to `Client().post`

RyanSkonnord committed:
Checks args presence; clean up assertions

RyanSkonnord committed:
Fix assertion in test_with_invalid_arguments

The previous assertion was wrong because the service wasn't correctly
raising RpcArgumentException, which it now is.

RyanSkonnord committed:
Merge branch 'master' into zc/rpc-in-tests

RyanSkonnord committed:
Make IntegrationService.send_incident_alert_notification more permissive

Having `Mapping[str, Any]` as an annotation is incompatible with our
long-term goals of converting the service signatures into a publishable
schema document, and invites brittle failure modes. We should try to
change this before locking in the schema.

corps committed:
More fixes

RyanSkonnord committed:
Debug user_service.serialize_many call in UserAvatarEndpoint

corps committed:
More fixes

RyanSkonnord committed:
Fix import glitch in SentryAppInstallationForProviderManager (fixes ProjectsListTest)

corps committed:
More fixes

RyanSkonnord committed:
Kludge comparisons in test_client_config_in_silo_modes

corps committed:
Fixes

RyanSkonnord committed:
Merge branch 'master' into zc/rpc-in-tests

RyanSkonnord committed:
Fix type-checking errors

getsantry[bot] committed:
:knife: regenerate mypy module blocklist

RyanSkonnord committed:
Fix IntegrationService.get_integration sig to match impl

RyanSkonnord committed:
Merge branch 'master' into zc/rpc-in-tests

RyanSkonnord committed:
In TeamManagerTest, use factories to set up user

Just a heuristic guess that this might fix errors that happen only on
the full CI run.

corps committed:
Yup

corps committed:
Merge branch 'master' into zc/rpc-in-tests

RyanSkonnord committed:
Merge branch 'master' into zc/rpc-in-tests

RyanSkonnord committed:
Serialize user before user_service.serialize_many call in ActivitySerializer

corps committed:
Merge branch 'master' into zc/rpc-in-tests

RyanSkonnord committed:
Merge branch 'master' into zc/rpc-in-tests

RyanSkonnord committed:
Clear env.request_stack with request

corps committed:
Pivot off silo mode acceptance tests

corps committed:
Merge branch 'master' into zc/rpc-in-tests

asottile-sentry reviewed (commented):
I just looked at the file count and there's way too much to review in this PR.  I did not do a thorough review because this is way too large

please split it into the functional parts

corps committed:
Fixing validations

corps committed:
More minor fixes

GabeVillalobos commented on src/sentry/models/integrations/integration.py:
```diff
@@ -106,18 +106,23 @@ def outboxes_for_update(identifier: int) -> List[ControlOutbox]:
 
-    def add_organization(self, organization: RpcOrganization, user=None, default_auth_id=None):
+    def add_organization(
+        self, organization_id: int | RpcOrganization, user=None, default_auth_id=None
```
nit: I think it'd be cleaner to just always have the caller supply the organization ID instead of passing in the full object if possible.

GabeVillalobos commented on src/sentry/services/hybrid_cloud/integration/impl.py:
```diff
@@ -343,7 +342,7 @@ def send_incident_alert_notification(
         organization: RpcOrganizationSummary,
         new_status: int,
-        incident_attachment: Mapping[str, str],
+        incident_attachment: Mapping[str, Any],
```
Why loosen the type here?

RyanSkonnord replied:
Just aligning with reality. The values returned by `sentry.rules.actions.notify_event_service.build_incident_attachment` have always had a nested dict as one of the values.

I've got a TODO in `service.py` to change the RPC interface -- an `Any` will be a no-go if we want to publish the schema.

GabeVillalobos commented on src/sentry/services/hybrid_cloud/organization/impl.py:
```diff
@@ -245,7 +245,7 @@ def set_user_for_organization_member(
 
     def check_organization_by_slug(self, *, slug: str, only_visible: bool) -> Optional[int]:
         try:
-            org = Organization.objects.get_from_cache(slug=slug)
```
I'm curious about this change. Was this just incorrect behavior before?

GabeVillalobos commented on src/sentry/silo/base.py:
```diff
@@ -43,7 +47,9 @@ def single_process_silo_mode(cls) -> bool:
     @contextlib.contextmanager
-    def enter_single_process_silo_context(cls, mode: SiloMode) -> Generator[None, None, None]:
+    def enter_single_process_silo_context(
+        cls, mode: SiloMode, region: Region | None = None
```
python question: If type-checking is false, does this just result in the type being deduced as `Any`?

GabeVillalobos commented on src/sentry/testutils/pytest/sentry.py:
```diff
@@ -219,6 +219,9 @@ def pytest_configure(config):
     settings.KAFKA_CONSUMER_FORCE_DISABLE_MULTIPROCESSING = True
 
+    # Assume this is always configured (not the real secret)
+    settings.RPC_SHARED_SECRET = ("215b1f0d",)
```
Do we have any enforcement mechanisms to ensure these are actually set in prod?

GabeVillalobos commented on tests/acceptance/sentry_plugins/test_amazon_sqs.py:
```diff
@@ -1,8 +1,8 @@
 
 
-@region_silo_test(stable=True)
+@no_silo_test(stable=True)
```
What does it mean to annotate a test with `no_silo_test` now? Just that tests are broken unless running in monolith mode?

corps replied:
Yes, and or more specifically that it is fine not to invoke the full RPC stack.

GabeVillalobos commented on tests/sentry/incidents/endpoints/test_organization_alert_rule_details.py:
```diff
@@ -172,7 +172,7 @@ def test_with_unresponsive_sentryapp(self):
             resp = self.get_response(self.organization.slug, self.rule.id)
 
-        assert len(responses.calls) == 1
+        # assert len(responses.calls) == 1
```
Should we just delete this now, or is there a better assertion we could be making here?

markstory commented on src/sentry/services/hybrid_cloud/organization/impl.py:
```diff
@@ -245,7 +245,7 @@ def set_user_for_organization_member(
     def check_organization_by_slug(self, *, slug: str, only_visible: bool) -> Optional[int]:
         try:
-            org = Organization.objects.get_from_cache(slug=slug)
+            org = Organization.objects.get(slug=slug)
```
Why no cache here? Seems like a good place to leverage caching.

corps replied:
I was debugging an issue, I can return that.

markstory commented on src/sentry/services/hybrid_cloud/rpc.py:
```diff
@@ -323,14 +342,9 @@ def _create_signatures(cls) -> Mapping[str, RpcMethodSignature]:
-                # While remote services are under development, swallow these errors
-                # and leave empty spots in the parameter model table. This will cause
-                # an RpcServiceUnimplementedException when the method is called.
-                # TODO: Make this a hard failure when all parameter models are stable
```
:clap: Nice to see these TODOs resolving.

markstory commented on src/sentry/services/hybrid_cloud/rpc.py:
```diff
@@ -398,31 +413,12 @@ def remote_method(service_obj: RpcService, **kwargs: Any) -> Any:
-                method = getattr(service, method_name)
-                return method(**kwargs)
+                serial_arguments = signature.serialize_arguments(kwargs)
+                return dispatch_remote_call(
+                    region, cls.key, method_name, serial_arguments, use_test_client=use_test_client
+                )
```
:clap: No more fallbacks :clap:

markstory commented on src/sentry/services/hybrid_cloud/rpc.py:
```diff
@@ -504,60 +504,136 @@ def result_to_dict(value: Any) -> Any:
+        with SiloMode.exit_single_process_silo_context(), SiloMode.enter_single_process_silo_context(
+            target_mode, self.region
+        ):
+            content_type = f"application/json; charset={_RPC_CONTENT_CHARSET}"
```
You could read the the content type from the provided headers
```suggestion
            content_type = headers.get('Content-Type')
```

🤷

corps committed:
Merge branch 'master' into zc/rpc-in-tests

markstory commented on tests/sentry/manager/test_team_manager.py:
```diff
@@ -1,13 +1,17 @@
 @region_silo_test
 class TeamManagerTest(TestCase):
+    def setUp(self) -> None:
+        env.clear()
```
Should you call `super.setUp()` here?

RyanSkonnord replied:
I believe it's a no-op but, yeah, I'll put that in for hygiene's sake.

RyanSkonnord committed:
Add super call to TeamManagerTest.setUp

corps commented on src/sentry/api/endpoints/rpc.py:
```diff
@@ -28,14 +33,37 @@ def post(self, request: Request, service_name: str, method_name: str) -> Respons
+        auth_context: AuthenticationContext = AuthenticationContext()
+        if auth_context_json := arguments.get("auth_context"):
+            try:
+                # Note -- generally, this is NOT set, but only in cases where an RPC needs to invoke code
+                # that depends on the `env.request.user` object.  In that case, the authentication context
+                # includes an authenticated user that will be injected into the global request context
+                # for compatibility.  Notably, this authentication context is *trusted* as the request comes
+                # from within the privileged RPC channel.
```
It's not possible generally though.  In a lot of places we are accessing this value inside of an ORM helper method that is easily 3-4 levels beyond the controller.  I think it will get messy fast.

corps commented on src/sentry/api/serializers/models/incidentactivity.py:
```diff
@@ -10,7 +11,8 @@ class IncidentActivitySerializer(Serializer):
         serialized_users = user_service.serialize_many(
-            filter={"user_ids": [i.user_id for i in item_list if i.user_id]}, as_user=user
+            filter={"user_ids": [i.user_id for i in item_list if i.user_id]},
+            as_user=serialize_generic_user(user),
```
In general I used `as_user` only when the original code used it.  🤷

corps committed:
Feedback

corps committed:
Typing fix

corps committed:
Merge branch 'master' into zc/rpc-in-tests

mrduncan commented on src/sentry/services/hybrid_cloud/app/impl.py:
```diff
@@ -126,12 +126,15 @@ def get_related_sentry_app_components(
-            group_by=group_by,
-        )
+        return {
+            str(k): v
```
Are there cases where `k` will be `None` (resulting in `"None"` instead of `None`)?

corps committed:
Feedback

corps committed:
Merge branch 'master' into zc/rpc-in-tests

corps committed:
Merge branch 'master' into zc/rpc-in-tests
